internals/daemon/api_notices.go

```diff
@@ -101,11 +118,63 @@ func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
 	return SyncResponse(notices)
 }
 
+// Get the UID of the request. If the UID is not known, return -1, indicating
+// that the connection may only receive notices intended for any recipient.
+func uidFromRequest(r *http.Request) int {
+	_, uid, _, err := ucrednetGet(r.RemoteAddr)
+	if err != nil {
+		return -1
+	}
+	return int(uid)
+}
+
+// Construct the user IDs filter which will be passed into the notices state.
+// Importantly, ensure that non-root users cannot filter on user IDs other than
+// their own and -1.
+func sanitizeUserIDsFilter(reqUid int, queryUserIDs []string) ([]int, error) {
+	userIDStrs := strutil.MultiCommaSeparatedList(queryUserIDs)
+	userIDs := make([]int, 0, len(userIDStrs))
+	for _, userIDStr := range userIDStrs {
+		uid, err := strconv.ParseInt(userIDStr, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		userID := int(uid)
+		if err := state.ValidateUserID(&userID); err != nil {
+			return nil, fmt.Errorf(`invalid user ID "%d": %v`, userID, err)
```

Comment:
No need for the quotes around %d here, as it's not a string:

```suggestion
			return nil, fmt.Errorf("invalid user ID %d: %v", userID, err)
```